src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -487,23 +490,69 @@ class RequiredMandatoryAttribute:
                         success = False
                         continue
 
+        self.print_step(3, "Validate the global attributes are in range and do not contain duplicates")
+        for endpoint_id, endpoint in self.endpoints_tlv.items():
+            for cluster_id, cluster in endpoint.items():
+                for req_attribute in ATTRIBUTES_TO_CHECK:
                     # Validate attribute value based on the provided validators.
                     for validator in req_attribute.validators:
                         try:
                             validator(cluster[req_attribute.id])
                         except ValueError as e:
+                            location = AttributePathLocation(endpoint_id, cluster_id, req_attribute.id)
                             self.record_error(self.get_test_name(), location=location,
                                               problem=f"Failed validation of value on {location.as_string(self.cluster_mapper)}: {str(e)}", spec_location="Global Elements")
                             success = False
                             continue
 
+        self.print_step(4, "Validate the attribute list exactly matches the set of reported attributes")
+        if success:
+            for endpoint_id, endpoint in self.endpoints_tlv.items():
+                for cluster_id, cluster in endpoint.items():
+                    attribute_list = cluster[ATTRIBUTE_LIST_ID]
+                    for attribute_id in attribute_list:
+                        location = AttributePathLocation(endpoint_id, cluster_id, attribute_id)
+                        has_attribute = attribute_id in cluster
+
+                        attribute_string = self.cluster_mapper.get_attribute_string(cluster_id, attribute_id)
+                        logging.debug(
+                            f"Checking presence of claimed supported {attribute_string} on {location.as_cluster_string(self.cluster_mapper)}: {'found' if has_attribute else 'not_found'}")
+
+                        # Check attribute is actually present.
+                        if not has_attribute:
+                            # TODO: Handle detecting write-only attributes from schema.
+                            if "WriteOnly" in attribute_string:
```

Comment:
Do we have examples of this in the current spec?

Reply:
This is not part of the new code - it's just moved. But as far as I know, we do not.